### PR TITLE
[RLlib] DreamerV3: Learner API classes for tf-keras, loss functions, additional update method.

### PR DESCRIPTION
**Pull request by sven1977**
Note: This PR is non-functional in itself. It merely introduces the individual losses and Learner API components (tf keras) needed to run the DreamerV3 algorithm, of which integration into RLlib is in progress (separate PRs will be released and merged in the next days).

DreamerV3:
* New Learner API classes for tf-keras
* All DreamerV3 loss functions in tf
* additional_update method for updating the critic's EMA copy.
* Makes enhancements on the Learner class' `compute_gradients`, `postprocess_gradients` and `apply_gradients` methods to be able to return stats dicts. These are then merged and returned from `Learner.update()`.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/torch/torch_learner.py [hunk 1]
```diff
@@ -84,46 +85,37 @@ def configure_optimizer_per_module(
...
             grad_clip=self._optimizer_config.get("grad_clip"),
-            grad_clip_by=self._optimizer_config.get("grad_clip_by"),
+            grad_clip_by=clip_by,
         )
+        stats = {}
+        if clip_by == "global_norm":
+            stats["gradients_global_norm"] = global_norm
 
-        return gradients_dict
+        return gradients_dict, stats
```
Comment (kouroshHakha) [inline rllib/core/learner/torch/torch_learner.py:115]: I see why you need to return the stats. My proposal is that instead of returning stats and results dict in general (changing the API) we should define a better stats / logging API for the learner, in particular:

```

def __init__():
      self._metrics = ...

def record_metric(key, value):
      self._metrics[key] = value

def record_metric_dict(metric_dict):
      self._metrics.update(**metric_dict)
       
```

Then you can do:
```
self.record_metric("gradirent_global_norm", global_norm)
```

**Review by kouroshHakha — COMMENTED** [review]
Can split this PR into two pieces? 

We need to settle on the requirements for the learner. I think there is so many changes proposed to the learner at the same time, it is making the learner API complex again. 

Read my suggestion from https://github.com/ray-project/ray/pull/35385#discussion_r1195942859  We can define a simple and yet effective stats reporting API inside the base learner class. `self.record_metrics/stats(key, value)`. With this you can get rid of all the stats dicts that you have to return. We will also define a getter method (`self.get_metrics(filter_fn=None)`) which will allow people to get metrics and apply custom filtering function on the fly. 

Also it makes the API so much nicer and cleaner. Let's do that in a separate PR first and then build the DreamerLearner API on top of it.

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+            learning_rate=self.hps.world_model_lr, epsilon=1e-8
+        )
+        optim_world_model.build(dreamerv3_module.world_model.trainable_variables)
+        ret["world_model"] = (
+            self.get_parameters(dreamerv3_module.world_model),
+            optim_world_model,
+        )
+        # Actor optimizer.
+        optim_actor = tf.keras.optimizers.Adam(
+            learning_rate=self.hps.actor_lr, epsilon=1e-5
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:64]: No clue, but the author's implementation uses these different epsilons. :) I honestly don't think it matters much, but haven't confirmed due to lack of time and resources.

File: rllib/core/learner/torch/torch_learner.py (hunk 1 above)
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py:115]: That's a great idea! Would simplify things again a bit. I was also worried about the complex new return structure, but didn't see a way to avoid duplicate computations, e.g. for the global norm, so we do need some way of getting these metrics out of the different gradient methods.
I'll change this.

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+        optim_critic.build(dreamerv3_module.critic.trainable_variables)
+        ret["critic"] = (
+            self.get_parameters(dreamerv3_module.critic),
+            optim_critic,
+        )
+
+        return ret
+
+    @override(TfLearner)
+    def compute_gradients(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:84]: Ah, I see, so we could simply sum up all 3 individual losses (world model, actor, critic) in one TOTAL_LOSS term and have the named optimizers (which already know their "parameter assignments") figure this out automatically with the base implementation's loop. Makes sense.

File: rllib/algorithms/dreamerv3/dreamerv3_learner.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from ray.rllib.core.learner.learner import Learner, LearnerHyperparameters
+from ray.rllib.core.rl_module.rl_module import ModuleID
+from ray.rllib.utils.annotations import override
+
+
+@dataclass
+class DreamerV3Hyperparameters(LearnerHyperparameters):
+    """Hyperparameters for the DreamerV3Learner sub-classes (framework specific).
+
+    These should never be set directly by the user. Instead, use the PPOConfig
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3_learner.py:22]: fixed

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py]: So the new way now is:
As soon as user needs more than one optimizer per module (and/or optimizers require rare c'tor settings), they must override `configure_optimizers_for_module`, which we do now here.

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+    "actor", and "critic". Each of these optimizers might ue a different learning rate,
+    epsilon parameter, and gradient clipping thresholds and procedures.
+    """
+
+    @override(TfLearner)
+    def configure_optimizer_per_module(
+        self, module_id: ModuleID
+    ) -> NamedParamOptimizerPairs:
+        """Create the 3 optimizers for Dreamer learning: world_model, actor, critic."""
+        ret: NamedParamOptimizerPairs = {}
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py]: fixed

File: rllib/core/learner/torch/torch_learner.py [hunk 6]
```diff
@@ -84,46 +85,37 @@ def configure_optimizer_per_module(
     @override(Learner)
     def compute_gradients(
         self, loss: Union[TensorType, Mapping[str, Any]]
-    ) -> ParamDictType:
+    ) -> Tuple[ParamDictType, ResultDict]:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py]: fixed (via new metrics API).

`compute_gradients`, `postprocess_gradients`, `apply_gradients` all DON'T return extra stats dicts. Users should simply register their metrics via the new API.

File: rllib/core/learner/torch/torch_learner.py [hunk 7]
```diff
@@ -136,6 +128,21 @@ def apply_gradients(self, gradients: ParamDictType) -> None:
...
+        self, module_id: ModuleID, *, timestep: int, **kwargs
+    ) -> Mapping[str, Any]:
+        results = super().additional_update_per_module(module_id, timestep=timestep)
+
+        # Handle lr scheduling updates and apply new learning rates to the optimizers.
+        new_lr = self.lr_scheduler.update(module_id=module_id, timestep=timestep)
+        results.update({LEARNER_RESULTS_CURR_LR_KEY: new_lr})
+
+        return results
+
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py]: Moved all this logic into base-class.

File: rllib/core/learner/torch/torch_learner.py (hunk 1 above)
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py]: done

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 8]
```diff
@@ -0,0 +1,719 @@
...
+
+        # Learn to produce symlog'd observation predictions.
+        # If symlog is disabled (e.g. for uint8 image inputs), `obs_symlog_BxT` is the
+        # same as `obs_BxT`.
+        obs_BxT = fwd_out["sampled_obs_symlog_BxT"]
+        obs_distr = fwd_out["obs_distribution_BxT"]
+        # Fold time dim and flatten all other (image?) dims.
+        obs_BxT = tf.reshape(
+            obs_BxT, shape=[-1, int(np.prod(obs_BxT.shape.as_list()[1:]))]
+        )
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:295]: Would this also work:
obs_BxT = tf.reshape(obs_BxT, shape=[-1, tf.reduce_prod(obs_BxT.shape[1:])])
?

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 9]
```diff
@@ -0,0 +1,719 @@
...
+            "L_prediction": L_pred,
+            "L_prediction_B_T": L_pred_B_T,
+        }
+
+    def _compute_world_model_dynamics_and_representation_loss(
+        self, *, hps, fwd_out
+    ) -> Tuple[TensorType, TensorType]:
+        """Helper method computing the world-model's dynamics and representation losses.
+
+        Dynamics loss: Trains the prior network, predicting z^ prior states from
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:369]: Returns:
    ....

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 10]
```diff
@@ -0,0 +1,719 @@
...
+
+    def _compute_value_targets(
+        self,
+        *,
+        hps: DreamerV3LearnerHyperparameters,
+        rewards_t0_to_H_BxT,
+        intrinsic_rewards_t1_to_H_BxT,
+        continues_t0_to_H_BxT,
+        value_predictions_t0_to_H_BxT,
+    ):
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:630]: Missing return type annotation

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 11]
```diff
@@ -0,0 +1,719 @@
...
+        return targets
+
+    def _compute_scaled_value_targets(
+        self,
+        *,
+        module_id,
+        hps: DreamerV3LearnerHyperparameters,
+        value_targets,
+        value_predictions,
+    ):
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:682]: Missing return type annotation

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Cool beans Sven 🙂 🎆

**Comment by ArturNiederfahrenhorst** [comment]
Some methods are lacking standardized docstrings with Args and Returns.

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 12]
```diff
@@ -0,0 +1,719 @@
...
+    ) -> NamedParamOptimizerPairs:
+
+        """Create the 3 optimizers for Dreamer learning: world_model, actor, critic."""
+        named_param_optimizer_pairs: NamedParamOptimizerPairs = {}
+
+        dreamerv3_module = self._module[module_id]
+
+        # World Model optimizer.
+        optim_world_model = tf.keras.optimizers.Adam(
+            learning_rate=hps.world_model_lr, epsilon=1e-8
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:56]: True, but I wanted to make at least the most obvious HP choices configurable. But in principal, you are right. Users should not really touch the config. But we'll see how well this will really work on envs other than those described in the paper. We still need to generate more prove that it's really a non-tuning-requiring algo :)

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py [hunk 13]
```diff
@@ -0,0 +1,719 @@
...
+
+
+class DreamerV3TfLearner(DreamerV3Learner, TfLearner):
+    """Implements DreamerV3 losses and gradient-based update logic in TensorFlow.
+
+    The critic EMA-copy update step can be found in the `DreamerV3Learner` base class,
+    as it is framework independent.
+
+    We define 3 local tensorflow optimizers for the sub components "world_model",
+    "actor", and "critic". Each of these optimizers might ue a different learning rate,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:40]: fixed

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py (hunk 8 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:295]: Yeah, let's try this. Seems cleaner :)

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py (hunk 9 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:369]: fixed

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py (hunk 10 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:630]: done

File: rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py (hunk 11 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/dreamerv3_tf_learner.py:682]: fixed

**Comment by ollie-iterators** [comment]
Lint error:
rllib/utils/replay_buffers/episode_replay_buffer.py:7:1: F401 'ray.rllib.policy.sample_batch.MultiAgentBatch' imported but unused
